Svc/FileDownlink/FileDownlink.cpp

```diff
@@ -127,106 +228,158 @@ namespace Svc {
+      this->cmdResponse_out(opCode, cmdSeq, Fw::COMMAND_EXECUTION_ERROR);
     }
+  }
 
+  void FileDownlink ::
+    Cancel_cmdHandler(
+        const FwOpcodeType opCode,
+        const U32 cmdSeq
+    )
+  {
+      //Must be able to cancel in both downlink and waiting states
+      if (this->mode.get() == Mode::DOWNLINK || this->mode.get() == Mode::WAIT) {
+          this->mode.set(Mode::CANCEL);
+      }
+      this->cmdResponse_out(opCode, cmdSeq, Fw::COMMAND_OK);
+  }
+
+  // ----------------------------------------------------------------------
+  // Private helper methods 
+  // ----------------------------------------------------------------------
+
+  Fw::CommandResponse FileDownlink :: 
+    statusToCmdResp(SendFileStatus status)
+  {
+    switch(status.e) {
+    case SendFileStatus::OK:
+      return Fw::COMMAND_OK;
+    case SendFileStatus::ERROR:
+      return Fw::COMMAND_EXECUTION_ERROR;
+    case SendFileStatus::INVALID:
+      return Fw::COMMAND_VALIDATION_ERROR;
+    case SendFileStatus::BUSY:
+        return Fw::COMMAND_BUSY;
+    default:
+        // Trigger assertion if given unknown status
+        FW_ASSERT(false);
+    }
+
+    // It's impossible to reach this, but added to suppress gcc missing return warning
+    return Fw::COMMAND_EXECUTION_ERROR;
+  }
+
+  void FileDownlink ::
+    sendResponse(SendFileStatus resp)
+  {
+    if(this->curEntry.source == FileDownlink::COMMAND) {
+      this->cmdResponse_out(this->curEntry.opCode, this->curEntry.cmdSeq, statusToCmdResp(resp));
+    } else {
+      for(NATIVE_INT_TYPE i = 0; i < this->getNum_FileComplete_OutputPorts(); i++) {
+        if(this->isConnected_FileComplete_OutputPort(i)) {
+          this->FileComplete_out(i, resp, this->curEntry.context, filenameString((char *) &this->curEntry.srcFilename));
+        }
+      }
+    }
+  }
+
+  void FileDownlink ::
+    sendFile(
+          const char* sourceFilename,
+          const char* destFilename,
+          U32 startOffset,
+          U32 length
+      )
+  {
     // Open file for downlink
-    status = this->file.open(
-        sourceFileName.toChar(),
-        destFileName.toChar()
+    Os::File::Status status = this->file.open(
+        sourceFilename,
+        destFilename
     );
 
     // Reject command if error when opening file
     if (status != Os::File::OP_OK) {
       this->mode.set(Mode::IDLE);
       this->warnings.fileOpenError();
       if (FILEDOWNLINK_COMMAND_FAIL_ON_MISSING_FILE) {
-          this->cmdResponse_out(opCode, cmdSeq, Fw::COMMAND_EXECUTION_ERROR);
+          sendResponse(SendFileStatus::ERROR);
       } else {
-          this->cmdResponse_out(opCode, cmdSeq, Fw::COMMAND_OK);
+          sendResponse(SendFileStatus::OK);
       }
       return;
     }
 
     // If the amount to downlink is greater than the file size, emit a Warning and then allow
     // the file to be downlinked anyway
     if (startOffset >= this->file.size) {
-        this->enterIdle();
+        this->enterCooldown();
```

Comment:
I think we can skip cool-down if this file failed to start...as we likely just had one.